src/core/lib/iomgr/timer_generic.c

```diff
@@ -79,6 +79,121 @@ static timer_shard g_shards[NUM_SHARDS];
  * Access to this is protected by g_shared_mutables.mu */
 static timer_shard *g_shard_queue[NUM_SHARDS];
 
+#ifndef NDEBUG
+
+/* == Hash table for duplicate timer detection == */
+
+#define NUM_HASH_BUCKETS 1000
+#define NUM_SLOTS_PER_BUCKET 30
+
+static gpr_mu g_hash_mu[NUM_HASH_BUCKETS]; /* One mutex per bucket */
+static grpc_timer *g_timer_ht[NUM_HASH_BUCKETS]
+                             [NUM_SLOTS_PER_BUCKET] = {{NULL, NULL}};
+
+static void init_timer_ht() {
+  for (int i = 0; i < NUM_HASH_BUCKETS; i++) {
+    gpr_mu_init(&g_hash_mu[i]);
+  }
+}
+
+static bool is_in_ht(grpc_timer *t) {
+  size_t i = GPR_HASH_POINTER(t, NUM_HASH_BUCKETS);
+  bool is_found = false;
+  gpr_mu_lock(&g_hash_mu[i]);
+  for (int j = 0; j < NUM_SLOTS_PER_BUCKET; j++) {
+    if (g_timer_ht[i][j] == t) {
+      is_found = true;
+      break;
+    }
+  }
+  gpr_mu_unlock(&g_hash_mu[i]);
+  return is_found;
+}
+
+static void add_to_ht(grpc_timer *t) {
+  size_t i = GPR_HASH_POINTER(t, NUM_HASH_BUCKETS);
+  bool added = false;
+  gpr_mu_lock(&g_hash_mu[i]);
+  for (int j = 0; j < NUM_SLOTS_PER_BUCKET; j++) {
+    if (g_timer_ht[i][j] == NULL) {
+      g_timer_ht[i][j] = t;
+      added = true;
+      break;
+    } else if (g_timer_ht[i][j] == t) {
+      grpc_closure *c = t->closure;
+      gpr_log(GPR_ERROR,
+              "** Duplicate timer (%p) being added. Closure: (%p), created at: "
+              "(%s:%d), scheduled at: (%s:%d) **",
+              t, c, c->file_created, c->line_created, c->file_initiated,
+              c->line_initiated);
+      abort();
+    }
+  }
+
+  gpr_mu_unlock(&g_hash_mu[i]);
+  if (!added) {
+    gpr_log(GPR_ERROR,
```

Comment:
How much more complicated would this be if we added a hash_table_next field to grpc_timer?

Reply:
(would allow chaining, and eliminate this class of failure)

Reply:
Hmm..actually chaining might be better (the reason for adding fixed bucket width was because I was initially curious about the max collisions we would get. Now I know that it is somewhere between 5 to 10 - and that too just for a few tests. Yeah, I will change this to chaining based).